### ISSUE TITLE
Allow sharing variable values between fragment, vertex and light shaders

### ISSUE THREAD

**Issue by ca3games** [godotengine/godot-proposals#630]
**Describe the project you are working on:**
A toon shader 3D demo.

**Describe the problem or limitation you are having in your project:**
I need to make a basic toon shader, but I can't use the light constant value from the light shader and use it to calculate the albedo in the fragment shader.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Have some variable or function that allow values to be shared between the vertex, or at least between the fragment and light shaders.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Have some variable type value you can pass between the fragment and light functions.
Or have some function to send information between the two functions.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
There's not way to store a value inside the light function and use it on the fragment shader (like the light vector).
Constants can't be used because I need the value in real time.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Not possible to do with an asset.

**Comment by groud** [comment]
This already exist between the vertes and the fragment shader: https://docs.godotengine.org/en/3.0/tutorials/shading/shading_language.html#varyings

I don't know how to do it between fragment and light though.

**Comment by clayjohn** [comment]
This comes up relatively frequently. Unfortunately, what you likely want is impossible due to the way shaders are executed on hardware. 

The vertex, fragment, and light functions are not run in arbitrary order, they have a precise orders, vertex, then fragment, then light. Each stage relies on the outputs of the previous stage. Accordingly, the fragment cannot send info to the vertex function and the light function cannot send info to the fragment function. 

Information can be passed the other way. You can send info from vertex to fragment using varyings. In theory it could be possible to create something to send info from the fragment to the light function (internally the light function is just a function called after the fragment function inside the fragment shader), but it would be extremely complex and likely of little benefit for most users. 

**TLDR;** You cant send information from the light function to the fragment function it is impossible. From fragment to light is possible in theory, but extremely impractical.

**Comment by Calinou** [comment]
Closing per @clayjohn's comment. The Vulkan renderer will likely not make this possible to implement either.
